Extend Cache Policies for the x86 Architecture

Tangzh33 opened the pull request:
This PR aims to improve the display efficiency of the EFI framebuffer. Previously, Asterinas supported only 4 cache policies: Writeback, Writethrough, Uncacheable minus, and Uncacheable. Built on these basic cache policies, the GUI has catastrophic display efficiency: it updates the screen line-by-line when a large area is refreshed.

Therefore, this PR supports advanced cache policies like Write Combining and Write Protected. Write Combining is suitable for the framebuffer. Here is some reference from the Intel x86 architecture manual.

> Write Combining (WC) — System memory locations are not cached (as with uncacheable memory) and coherency is not enforced by the processor’s bus coherency protocol. Speculative reads are allowed. Writes may be delayed and combined in the write combining buffer (WC buffer) to reduce memory accesses.
> 

> This type of cache-control is appropriate for video frame buffers, where the order of writes is unimportant as long as the writes update memory so they can be seen on the graphics display.
> 

Additionally, most mainstream Intel x86 CPUs support the feature, so I think it is appropriate to include the advanced cache policy in the main branch rather than keep it as an experimental implementation.

> This memory type is available in the Pentium Pro and Pentium II processors by programming the MTRRs; or in processor families starting from the Pentium III processors by programming the MTRRs or by selecting it through the PAT.
> 

![x86PageTableLayout](https://github.com/user-attachments/assets/2683d551-82bc-4ce0-878d-f74872daecab)

The implementation has already been tested on the experimental branch with framebuffer GUI support. And the display efficiency is much better. However, I have two open questions and do not have specific answers yet:

1. How should we incorporate the customized cache policy in our device memory framework? It is reasonable that our current device memory is populated when `mmap` is used, since device memory does not need to support on-demand paging. However, the cache policy for mapping `IoMem` is not customizable; by default, it is Uncachable (`VmMapping::populate_device`). We should support `IoMem` cache policy customization by providing more context for the device to use when customizing, and for the `mmap` logic to use.
2. How should we test the new cache policy with CI? The current implementation has no compilation errors and has proven effective on Xfce. But efficiency can only be validated by the eye. Can we incorporate a formal test to validate?

With questions to be decided, this PR is temporarily a draft PR, and will be ready when the questions are solved :-)

cc @junyang-zh @lrh2000

junyang-zh commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -69,6 +68,12 @@ bitflags::bitflags! {
+        /// Valid page marker. It indicates a valid leaf page entry at level 1.
+        /// For levels 2 and 3, the `HUGE` bit indicates a valid leaf entry. For
+        /// levels above 3, `PRESENT` is used for "valid".
+        const VALID_PAGE =      1 << 55;
```
The comments could have mentioned that it is ignored by the hardware. We use it for validity tracking.

junyang-zh commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -214,13 +230,35 @@ impl PageTableEntryTrait for PageTableEntry {
+        //  1   0   0  -> WC (Write Combining)
+        //  1   0   1  -> WP (Write Protected)
+        //  1   1   0  -> UC (Uncacheable)
+        //  1   1   1  -> UC (Uncacheable)
```
This looks ugly and prone to error if we will change the table later. And, the programmer's guide suggests clearly that you should combine the `PAT PCD PWT` bits as an 3-bit integer index to the array. So we should do so.

Firstly, map cache policies to PAT entry values instead of using your `const PAT_ENTRIES: [u64; 8]` array.

```rust
impl CachePolicy {
    /// Reference: Table 11-10 in Intel SDM Vol. 3A.
    fn to_ia32_pat_memory_type_entry(&self) -> u8 { match self { /* ... */ } }
}
```

Then, define our software-defined policy mappings.

```
const IA32_PAT_MAPPINGS: [CachePolicy; 8] = [/*...*/];
```

To query the policy, this array could be indexed by the 3-bit integer here.

Finally, when doing `configure_pat`, map each values in this array.

junyang-zh commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -302,3 +379,33 @@ impl fmt::Debug for PageTableEntry {
+    //   6: Uncacheable (UC)
+    //   7: Uncacheable (UC)
+
+    unsafe {
```
Always write safety reasoning. Just explain what it means to read/write these MSRs.

lrh2000 commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -270,21 +308,60 @@ impl PageTableEntryTrait for PageTableEntry {
+        // A PTE is "last" (leaf mapping) if:
+        // 1. It has HUGE bit set (huge page at level 2/3), OR
+        // 2. It has VALID_PAGE bit set (4KB page at level 1)
+        self.0 & PageTableFlags::HUGE.bits() != 0 || self.0 & PageTableFlags::VALID_PAGE.bits() != 0
```
We do we want to do this instead of checking `level == 1 || self.0.contains(PageTableFlags::HUGE)`?

https://github.com/asterinas/asterinas/blob/0615e6f6424ead8e12d73670428149fa7270571f/ostd/src/arch/loongarch/mm/mod.rs#L289-L291

This can simplify the related logics a lot.

junyang-zh commented:
> > I see no problem in the approach of using another unused bit instead of PAT to test validity.
>
> The problem may be that it will cause confusion. So I think it needs to be justified why it's better than the other alternatives (see some random comments above).

The thing is we still need a bit to mark valid a page with no protection permissions (not readable, so in x86 no `PRESENT` bit). I agree that `is_last` needn't consult this bit.

Tangzh33 commented:
@tatetian @junyang-zh @lrh2000 Thanks for your insightful discussions, and I have refactored the code accordingly.

In this iteration, I adopted @lrh2000's suggestion and replaced the `Uncacheable Minus` with `Write Combining`. The changes look clean and straightforward, as expected. Additionally, I enhanced `IoMem` to support customizable cache policies and updated the related code accordingly. Since the main open questions have been thoroughly discussed and reached consensus, this PR is no longer a draft and is ready for review. Please leave your comments at your convenience~

junyang-zh commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -160,6 +181,10 @@ impl PageTableEntry {
+
+    fn is_huge(&self) -> bool {
+        self.0 & PageTableFlags::HUGE.bits() != 0
+    }
```
No I don't think it can exist because it is way too ambiguous. We encode entries in different levels with the same struct, and in different levels the huge bit doesn't always mean the huge bit.

junyang-zh commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -214,13 +239,16 @@ impl PageTableEntryTrait for PageTableEntry {
+        // Determine cache policy from PAT, PCD, PWT bits.
+        // For leaf pages (level 1 4KB pages and level 2/3 huge pages), bit 7 (HUGE) is always 1,
+        // so they all use PAT indices 4-7, which mirror indices 0-3 in our configuration.
+        let pat_bit = self.is_huge();
```
This is simply wrong. On level-2/level-3 huge page mappings, the PAT bit is the 12th bit, not the huge bit. You can add a `PAT_LVL_2_3` bit, and use the `level` parameter to tell which to use (`PAT_FOR_HUGE` or `HUGE`).

junyang-zh commented on ostd/src/mm/page_prop.rs:
```diff
@@ -96,6 +96,22 @@ pub enum CachePolicy {
 
+#[cfg(target_arch = "x86_64")]
+impl CachePolicy {
+    /// Returns the IA32_PAT memory type encoding for this cache policy.
```
Be consistent on formatting. Since you wrap `IA32_PAT` with backticks erlier, do it here as well.

tatetian commented on ostd/src/io/io_mem/mod.rs:
```diff
@@ -172,12 +180,12 @@ impl IoMem<Sensitive> {
-    /// Acquires an `IoMem` instance for the given range.
-    pub fn acquire(range: Range<Paddr>) -> Result<IoMem<Insensitive>> {
+    /// Acquires an `IoMem` instance for the given range with the specified cache policy.
+    pub fn acquire(range: Range<Paddr>, cache_policy: CachePolicy) -> Result<IoMem<Insensitive>> {
```
`IoMem` is mainly introduced to represent MMIO, which normally use `Uncacheable`. Only graphics display buffers use `WriteCombining`. So I would say that in 90% situations, the users of `IoMem` choose `Uncacheable`. So a more user-friendly design is to have two separate APIs:

```rust
impl IoMem {
    pub fn acquire(range: Range<Paddr>) -> Result<IoMem<Insensitive>> {
        Self::acquire_with_policy(range, CachePolicy::Uncachable)
    }

    pub fn acquire_with_policy(range: Range<Paddr>, cache_policy: CachePolicy) -> Result<IoMem<Insensitive>> { ... }
}
```

junyang-zh replied:
IMHO `pub fn acquire(range: Range<Paddr>, cache: CachePolicy)` is fine.

tatetian replied:
**A user-friendly API design makes the common cases easy, the rare cases possible.** So I stand with my current position that there should be two constructor methods: one with a short name for the common case where the cache policy is `CachePolicy::Uncacheable`, the other with a longer name for the rare case where the cache policy needs to be given.

> Another problem is that it's unclear what kind of policy `acquire_with_policy` wants to mean.

To enhance clarity (on the price of a longer name), we can rename it to `acquire_with_cache_policy`. Since it is rarely used, the longer name is still acceptable.

So how about this one?

```rust
impl IoMem {
    pub fn acquire(range: Range<Paddr>) -> Result<IoMem<Insensitive>> {
        Self::acquire_with_cache_policy(range, CachePolicy::Uncachable)
    }

    pub fn acquire_with_cache_policy(range: Range<Paddr>, cache_policy: CachePolicy) -> Result<IoMem<Insensitive>> {
       // ..
    }
}
```

lrh2000 replied:
Okay. I still think `acquire_uncached` could be clearer for the aforementioned reason, but I'm also fine with the current version.

tatetian reviewed (commented):
Great work. The changes look mostly good.

Tangzh33 commented:
@tatetian @junyang-zh Thank you for your attention! Your comments have been addressed in this iteration. This PR is ready for review again; please let me know if you have any further comments :-)

lrh2000 commented:
> The thing is we still need a bit to mark valid a page with no protection permissions (not readable, so in x86 no `PRESENT` bit).

@junyang-zh So what do you think is the expected behavior if the underlying architecture-specific page table cannot support the required flags?

For example, what would happen if we tried to protect pages to be write-only? If we deny it (or ignore it), why not deny (or ignore) "not readable" as well?

A similar question is what will happen if we specify the cache policy that the underlying page table cannot support (e.g., RISC-V without `SVPBMT`).

Besides, we should not make use of `RWX=000` in RISC-V as well because it's reserved for future use per the spec?

lrh2000 commented on ostd/src/mm/page_prop.rs:
```diff
@@ -96,6 +96,22 @@ pub enum CachePolicy {
+    /// Returns the `IA32_PAT` memory type encoding for this cache policy.
+    ///
+    /// Reference: Intel 64 and IA-32 Architectures SDM Vol. 3A, Table 11-10.
+    pub(crate) const fn to_ia32_pat_memory_type_entry(self) -> u8 {
```
This should be moved to the x86 module? I suspect it should live inside `configure_pat`...
```rust
pub(crate) fn configure_pat() {
    fn cache_policy_to_pat_entry(cache_policy: CachePolicy) -> u8 { ... }
}
```

lrh2000 commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -302,3 +330,20 @@ impl fmt::Debug for PageTableEntry {
+    }
+
+    // SAFETY: Writing `IA32_PAT` merely programs the page attribute table MSR.
+    // This code runs on the bootstrap processor before other CPUs depend on
```
This is not true. The code will also run in APs.
---

To make things simpler, I think most operations in `enable_cpu_features` is safe just because it builds up our global invariant of the CPU state. We'd better figure out a better way to write safety comments.

lrh2000 commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -20,6 +21,22 @@ use crate::{
+    CachePolicy::Writethrough,   // Index 5: PAT=1, PCD=0, PWT=1 (same as 1)
+    CachePolicy::WriteCombining, // Index 6: PAT=1, PCD=1, PWT=0 (same as 2)
+    CachePolicy::Uncacheable,    // Index 7: PAT=1, PCD=1, PWT=1 (same as 3)
+];
```
```rust
const IA32_PAT_MAPPINGS: [CachePolicy; 8] = [
    CachePolicy::Writeback,      // Index 0: PAT=0, PCD=0, PWT=0
    CachePolicy::Writethrough,   // Index 1: PAT=0, PCD=0, PWT=1
    CachePolicy::WriteCombining, // Index 2: PAT=0, PCD=1, PWT=0 (replaces UC-)
    CachePolicy::Uncacheable,    // Index 3: PAT=0, PCD=1, PWT=1
    CachePolicy::Writeback,      // Index 4: PAT=1, PCD=0, PWT=0 (same as 0)
    CachePolicy::Writethrough,   // Index 5: PAT=1, PCD=0, PWT=1 (same as 1)
    CachePolicy::WriteCombining, // Index 6: PAT=1, PCD=1, PWT=0 (same as 2)
    CachePolicy::Uncacheable,    // Index 7: PAT=1, PCD=1, PWT=1 (same as 3)
];

const fn pat_to_cache_policy(page_flags: PageTableFlags) -> CachePolicy {
    let mut index = 0usize;
    if page_flags.contains(PageTableFlags::NO_CACHE) {
        index |= 2;
    }
    if page_flags.contains(PageTableFlags::WRITE_THROUGH) {
        index |= 1;
    }
    IA32_PAT_MAPPINGS[index]
}

const fn cache_policy_to_pat(cache_policy: CachePolicy) -> PageTableFlags {
    match cache_policy {
        CachePolicy::Writeback => PageTableFlags::empty(),
        CachePolicy::Writethrough => PageTableFlags::WRITE_THROUGH,
        CachePolicy::Uncacheable => PageTableFlags::NO_CACHE.union(PageTableFlags::WRITE_THROUGH),
        CachePolicy::WriteCombining => PageTableFlags::NO_CACHE,
        _ => panic!("unsupported cache policy"),
    }
}

const_assert!(matches!(
    pat_to_cache_policy(cache_policy_to_pat(CachePolicy::Writeback)),
    CachePolicy::Writeback
));
const_assert!(matches!(
    pat_to_cache_policy(cache_policy_to_pat(CachePolicy::Writethrough)),
    CachePolicy::Writethrough
));
const_assert!(matches!(
    pat_to_cache_policy(cache_policy_to_pat(CachePolicy::Uncacheable)),
    CachePolicy::Uncacheable
));
const_assert!(matches!(
    pat_to_cache_policy(cache_policy_to_pat(CachePolicy::WriteCombining)),
    CachePolicy::WriteCombining
));
```

 1. Define methods that perform the translation work, and make the methods live near to `IA32_PAT_MAPPINGS`.
 2. Use constant assertions to ensure the correctness.
 3. These methods appear too early in this file. Try to find a better position. Well, I think creating a new `pat.rs` may be a better idea. So they can live near to `configure_pat`.

junyang-zh commented:
> For example, in x86 fallback to RW for write-only

Acutally the status quo is, we will set `PRESENT=0 PAT/HUGE=1 WRITABLE=1` for write-only, and the MMU will fault on such writes. The actual fallback is done by the upper kernel...

It's kind of a mess right now.

lrh2000 commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -303,3 +304,46 @@ impl fmt::Debug for PageTableEntry {
+
+/// Programs the PAT MSR so that write-combining mappings use the
+/// correct memory type.
+fn configure_pat() {
```
1. Move this to `pat.rs`.
 2. Remove `pub(super)` from `IA32_PAT_MAPPINGS`.

lrh2000 commented on ostd/src/arch/x86/cpu/extension.rs:
```diff
@@ -112,6 +112,7 @@ define_isa_extensions! {
         RDRAND,       Ecx(30), "The processor supports RDRAND instruction.";
 
         XAPIC,        Edx( 9), "APIC On-Chip.";
+        PAT,          Edx(16), "Supports the Page Attribute Table (PAT).";
```
There is no need to define this, nor do I intend to. At least, we don't define "FPU" because it's available on all x86-64 CPUs.

lrh2000 commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -303,3 +304,46 @@ impl fmt::Debug for PageTableEntry {
+    // SAFETY: Writing `IA32_PAT` only programs the PAT MSR of the current CPU.
+    // Updating PAT merely redefines how hardware interprets future cache
+    // policy encodings. No system components rely on the contents of the PAT
+    // being modified before this point.
```
```rust
    // policy encodings. The programmed value is the global invariant, which
    // is set up before the kernel page table is activated.
```

lrh2000 commented on ostd/src/arch/x86/mm/mod.rs:
```diff
@@ -303,3 +304,46 @@ impl fmt::Debug for PageTableEntry {
+pub(super) fn enable_essential_features() {
+    // PAT has been available since Pentium III (1999) and is ubiquitous in
+    // modern 64-bit CPUs. Therefore, we assume that all x86-64 CPUs should
+    // have Page Attribute Table (PAT) support. Otherwise, we should check
```
https://github.com/asterinas/asterinas/pull/2570#discussion_r2525915985 means:
```rust
    // Page Attribute Table (PAT) has been available since Pentium III (1999)
    // and is ubiquitous in modern 64-bit CPUs. Therefore, we assume that all
    // x86-64 CPUs should have PAT support. Otherwise, we should check
```

lrh2000 commented on ostd/src/arch/x86/mm/pat.rs:
```diff
@@ -0,0 +1,64 @@
+
+pub(super) const fn cache_policy_to_flags(cache_policy: CachePolicy) -> PageTableFlags {
+    let bits = match cache_policy {
+        CachePolicy::Writeback => 0,
```
Keep the reference in the previous version.

lrh2000 replied:
Oh I'm sorry. I posted the comment on the wrong line, so it's confusing.

I mean the comment should be added back in line 71:
```rust
    // Reference: Intel(R) 64 and IA-32 Architectures Software Developer's Manual, Table 12-10,
    // "Memory Types That Can Be Encoded With PAT".
    fn cache_policy_to_pat_entry(cache_policy: CachePolicy) -> u8 {
        match cache_policy {
            CachePolicy::Uncacheable => 0x00,
            CachePolicy::WriteCombining => 0x01,
            CachePolicy::WriteProtected => 0x05,
            CachePolicy::Writethrough => 0x04,
            CachePolicy::Writeback => 0x06,
        }
    }
```

Tangzh33 replied:
Thanks for your reply. I have added the requested comments. This PR is now ready for review again. Please leave your comments at your convenience~

lrh2000 reviewed (commented):
Thanks! Just some minor comments.

tatetian commented:
@Tangzh33 Need to fix the conflict. Thanks!
